src/dataset_impl.cpp

```diff
@@ -29,4 +33,288 @@ DatasetImpl::MakeEmptyDataset() {
     return result;
 }
 
+template <typename T>
+T*
+allocate_and_copy(
+    const T* src, size_t count, Allocator* allocator, T* old_dest = nullptr, size_t old_count = 0) {
+    if (!src || count == 0) {
+        return nullptr;
+    }
+    if (old_dest == nullptr && old_count > 0) {
+        throw VsagException(ErrorType::INVALID_ARGUMENT,
+                            "Old destination cannot be null if old count is greater than zero");
+    }
+    if (old_dest && old_count == 0) {
+        throw VsagException(ErrorType::INVALID_ARGUMENT,
+                            "Old count must be greater than zero if old destination is provided");
+    }
+
+    T* dest;
+    if (allocator != nullptr) {
+        if (old_dest != nullptr) {
+            // If old_dest is provided, we need to reallocate new memory
+            dest =
+                static_cast<T*>(allocator->Reallocate(old_dest, (old_count + count) * sizeof(T)));
+        } else {
+            // Allocate new memory for the new count
+            dest = static_cast<T*>(allocator->Allocate(count * sizeof(T)));
+        }
+    } else {
+        if (old_dest != nullptr) {
+            // If old_dest is provided, we need to allocate new memory
+            dest = new T[old_count + count];
+            memcpy(dest, old_dest, old_count * sizeof(T));
+            delete[] old_dest;  // Free the old memory if it was allocated with new[]
+        } else {
+            dest = new T[count];
+        }
+    }
```

Comment:
try to extract the actions with an inline function, like:

```
inline void
extend_old_dest(void & old_dest, old_count, count);

inline void
use_allocator_to_allocate_new_dest(...);

inline void
use_new_to_allocate_new_and_copy_old(...);

inline void
use_new_to_allocate_new(...);
```

and change this if conditions:

```
if (allocator != nullptr and old_dest != nullptr) {
extend_old_dest(...);
} else if (allocator != nullptr and old_dest == nullptr) {
use_allocator_to_allocate_new_dest(...);
} else if (...) { ... }
```